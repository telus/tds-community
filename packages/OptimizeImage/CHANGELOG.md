# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [1.2.0](https://github.com/telus/tds-community/compare/@tds/community-optimize-image@1.1.0...@tds/community-optimize-image@1.2.0) (2021-09-21)


### Features

* **community-optimize-image:** react 17 and styled components 5 ([75755b0](https://github.com/telus/tds-community/commit/75755b0eb112986a1f8eadc298b2c4d23b8af300))





# [1.1.0](https://github.com/telus/tds-community/compare/@tds/community-optimize-image@1.0.1...@tds/community-optimize-image@1.1.0) (2021-06-10)


### Features

* **community-optimize-image:** adding backwards compatibility ([c6603ac](https://github.com/telus/tds-community/commit/c6603ac0b0c80bee99f6040d04fe4294774bf766))





## [1.0.1](https://github.com/telus/tds-community/compare/@tds/community-optimize-image@1.0.0...@tds/community-optimize-image@1.0.1) (2021-04-16)

**Note:** Version bump only for package @tds/community-optimize-image





# 1.0.0 (2021-04-15)


### Bug Fixes

* **community-optimize-image:** update optimize-image readme ([b9da197](https://github.com/telus/tds-community/commit/b9da19752e41d9d1d745c03fb3f33230ce0316a4))


### Features

* **community-optimize-image:** add checkWebpFeature ([65b6b0c](https://github.com/telus/tds-community/commit/65b6b0ce601646128a6e743aa382818384a28edb))
* **community-optimize-image:** add customizable widths, update docs ([d52e36a](https://github.com/telus/tds-community/commit/d52e36a811c1237e923a9fa0ba89a1cdfdf34290))
* **community-optimize-image:** init new component ([b8216df](https://github.com/telus/tds-community/commit/b8216df35939f73dff2f89707bde663319ea1444))
* **community-optimize-image:** replace comment ([a760c35](https://github.com/telus/tds-community/commit/a760c3571cf7e0b2e1a9c70bece5112ae589d98b))
