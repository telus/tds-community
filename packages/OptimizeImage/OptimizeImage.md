OptimizeImage is used to optimize image files based on screen size

### Usage criteria

- Must use a contentful asset url as the component adapts the url to convert the image file
- May customize quality and screen width sizes. Note: increasing quality and width may affect page performance.

```jsx
<OptimizeImage
  contentfulAssetUrl="https://images.ctfassets.net/3cqlnin176yn/1GfIHDOb3n3uO3wVnKjHZ2/3a3c59c5caa91a315da1b3804cdf0b1b/Alpaca-lets-make-the-future-friendly-Hero-Banner-Tile.jpg"
  alt="alpacas"
  lg={500}
/>
```
