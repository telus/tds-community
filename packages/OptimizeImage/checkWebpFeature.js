// taken directly from Google developers guide on how to detect browser support for WebP
export default callback => {
  // basic support
  const lossy = 'UklGRiIAAABXRUJQVlA4IBYAAAAwAQCdASoBAAEADsD+JaQAA3AAAAAA'
  // other test forms
  // var kTestImages = {
  //   lossy: "UklGRiIAAABXRUJQVlA4IBYAAAAwAQCdASoBAAEADsD+JaQAA3AAAAAA",
  //   lossless: "UklGRhoAAABXRUJQVlA4TA0AAAAvAAAAEAcQERGIiP4HAA==",
  //   alpha: "UklGRkoAAABXRUJQVlA4WAoAAAAQAAAAAAAAAAAAQUxQSAwAAAARBxAR/Q9ERP8DAABWUDggGAAAABQBAJ0BKgEAAQAAAP4AAA3AAP7mtQAAAA==",
  //   animation: "UklGRlIAAABXRUJQVlA4WAoAAAASAAAAAAAAAAAAQU5JTQYAAAD/////AABBTk1GJgAAAAAAAAAAAAAAAAAAAGQAAABWUDhMDQAAAC8AAAAQBxAREYiI/gcA"
  // };
  const img = new Image()
  img.onload = function onLoad() {
    const result = img.width > 0 && img.height > 0
    callback(result)
  }
  img.onerror = function onError() {
    callback(false)
  }
  img.src = `data:image/webp;base64,${lossy}`
}
