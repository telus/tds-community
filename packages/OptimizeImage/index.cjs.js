const OptimizeImage = require('./dist/index.cjs')

module.exports = OptimizeImage
