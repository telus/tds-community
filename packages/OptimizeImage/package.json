{
  "name": "@tds/community-optimize-image",
  "version": "1.2.0",
  "description": "",
  "main": "index.cjs.js",
  "module": "index.es.js",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/telus/tds-community.git"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "TELUS digital",
  "engines": {
    "node": ">=8"
  },
  "bugs": {
    "url": "https://github.com/telus/tds-community/issues"
  },
  "homepage": "http://tds.telus.com",
  "peerDependencies": {
    "react": "^18.0.0",
    "react-dom": "^18.0.0"
  },
  "dependencies": {
    "@tds/core-responsive-image": "^1.0.0",
    "@tds/util-helpers": "^1.5.0",
    "prop-types": "^15.6.2"
  }
}
