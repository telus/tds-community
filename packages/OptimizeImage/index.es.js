import OptimizeImage from './dist/index.es'

export default OptimizeImage
