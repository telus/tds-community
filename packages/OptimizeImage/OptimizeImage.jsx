import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import ResponsiveImage from '@tds/core-responsive-image'
import { breakpoints } from '@tds/core-responsive'
import { safeRest } from '@tds/util-helpers'
import checkWebpFeature from './checkWebpFeature'

/**
 * @version ./package.json
 */
const OptimizeImage = ({ contentfulAssetUrl, alt, quality, ...rest }) => {
  // states used to ensure images are downloaded only once
  const [hasLoaded, setHasLoaded] = useState(false)
  const [imgUrls, setImgUrls] = useState({})

  useEffect(() => {
    if (contentfulAssetUrl && !contentfulAssetUrl.match(/.svg/g)) {
      // Currently not all browsers support webP
      checkWebpFeature(result => {
        let format = ''
        if (result) {
          format = 'fm=webp'
        } else if (contentfulAssetUrl.match(/.jpeg/g)) {
          format = 'fm=jpg&fl=progressive'
        }

        setImgUrls({
          xsSrc: `${contentfulAssetUrl}?w=320&q=${quality}&${format}`,
          smSrc: `${contentfulAssetUrl}?w=${breakpoints.sm}&q=${quality}&${format}`,
          mdSrc: `${contentfulAssetUrl}?w=${breakpoints.md}&q=${quality}&${format}`,
          lgSrc: `${contentfulAssetUrl}?w=${breakpoints.lg}&q=${quality}&${format}`,
          xlSrc: `${contentfulAssetUrl}?w=${breakpoints.xl}&q=${quality}&${format}`,
          fallbackSrc: `${contentfulAssetUrl}?w=${breakpoints.xl}&q=${quality}`,
        })
        setHasLoaded(true)
      })
    }
  }, [])

  if (!hasLoaded) return null
  return (
    <ResponsiveImage
      xsSrc={imgUrls.xsSrc}
      smSrc={imgUrls.smSrc}
      mdSrc={imgUrls.mdSrc}
      lgSrc={imgUrls.lgSrc}
      xlSrc={imgUrls.xlSrc}
      fallbackSrc={imgUrls.fallbackSrc}
      alt={alt}
      {...safeRest(rest)}
    />
  )
}

OptimizeImage.propTypes = {
  /**
   * The source to load the image.
   */
  contentfulAssetUrl: PropTypes.string.isRequired,
  /**
   * Alternative text to display if image cannot be loaded or a screen reader is used.
   */
  alt: PropTypes.string.isRequired,
  /**
   * Customize quality as a percentage between 1 and 100.
   */
  quality: PropTypes.number,
}

OptimizeImage.defaultProps = {
  quality: 80,
}

export default OptimizeImage
