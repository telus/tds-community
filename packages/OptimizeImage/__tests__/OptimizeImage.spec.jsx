import React from 'react'
import { mount } from 'enzyme'

import OptimizeImage from '../OptimizeImage'
import checkWebpFeature from '../checkWebpFeature'

jest.mock('../checkWebpFeature')

describe('OptimizeImage', () => {
  beforeEach(() => {
    checkWebpFeature.mockImplementation(res => res(true))
  })

  const doMount = (props = {}) =>
    mount(
      <OptimizeImage
        contentfulAssetUrl="https://images.ctfassets.net/Alpaca.jpg"
        alt="test-image"
        {...props}
      />
    )

  it('renders', () => {
    const optimizeImage = doMount()
    expect(optimizeImage).toMatchSnapshot()
  })

  it('passes additional attributes to the element', () => {
    const optimizeImage = doMount({ id: 'the-id', 'data-some-attr': 'some value' })

    expect(optimizeImage).toHaveProp('id', 'the-id')
    expect(optimizeImage).toHaveProp('data-some-attr', 'some value')
  })

  it('does not allow custom CSS', () => {
    const optimizeImage = doMount({
      className: 'my-custom-class',
      style: { color: 'hotpink' },
    }).find('ResponsiveImage')

    expect(optimizeImage).not.toHaveProp('className', 'my-custom-class')
    expect(optimizeImage).not.toHaveProp('style')
  })

  it('should not alter the url if it is an svg', () => {
    const optimizeImage = doMount({
      contentfulAssetUrl: 'https://images.ctfassets.net/Alpaca.svg',
    }).find('ResponsiveImage')

    expect(optimizeImage).toHaveProp('xsSrc', 'https://images.ctfassets.net/Alpaca.svg')
  })

  it('should format with webp if browser supports it', () => {
    const optimizeImage = doMount({
      contentfulAssetUrl: 'https://images.ctfassets.net/Alpaca.jpg',
    }).find('ResponsiveImage')

    expect(optimizeImage).toHaveProp(
      'xsSrc',
      'https://images.ctfassets.net/Alpaca.jpg?w=320&q=80&fm=webp'
    )
  })

  it('should format with progressive jpg if browser does not support webp', () => {
    checkWebpFeature.mockImplementation(res => res(false))

    const optimizeImage = doMount({
      contentfulAssetUrl: 'https://images.ctfassets.net/Alpaca.jpeg',
    }).find('ResponsiveImage')

    expect(optimizeImage).toHaveProp(
      'xsSrc',
      'https://images.ctfassets.net/Alpaca.jpeg?w=320&q=80&fm=jpg&fl=progressive'
    )
  })
})
