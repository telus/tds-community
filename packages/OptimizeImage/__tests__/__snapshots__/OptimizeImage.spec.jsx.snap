// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptimizeImage renders 1`] = `""`;
